### PR TITLE
RxCpp and Disruptor integration

### PR DESCRIPTION
**Pull request by ivan-cukic**

### CODE REVIEW

**Review by wirew0rm — CHANGES_REQUESTED** [review]
Very nice work, the API feels really nice.

I left some comments on the cmake side. Since I have another PR which changes lots of things in the build system anyway, we should concentrate on the c++ side here, feel free to change things that are easy to fix and leave the rest for the cmake cleanup.

In the code I have some small things and a few questions about how things work.

<strike>On the API side I noticed that we hava a Source, for going from disruptor to Rx and wie have Aggregation which aggregates and translates back to the Disruptor. I was wondering if it would be more flexible to have Aggregation purely in Rx and then a separate Sink which republishes into the Disruptor. Or would this have a big impact on performance or API?</strike>
Misread the code, the aggregated events are not published into the Disruptor again, I was mislead by the way it is implemented in java here.

File: src/disruptor/include/disruptor/Exception.hpp [hunk 1]
```diff
@@ -13,16 +13,15 @@ class DisruptorException : public std::exception {};
...
 class WrappedException : public DisruptorException {
 private:
-    std::string           message;
-    const std::exception &inner_exception;
+    std::string message;
 
 public:
     explicit WrappedException(const std::exception &ex, std::string msg)
-        : message(std::move(msg)), inner_exception(ex) {
+        : message(std::move(msg)) {
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply src/disruptor/include/disruptor/Exception.hpp:20]: Ok, though storing a reference to an exception is potentially unsafe WRT lifetime.

File: CMakeLists.txt [hunk 2]
```diff
@@ -76,6 +76,7 @@ if (ENABLE_PCH)
             <utility>)
 endif ()
 
+set(CONAN_DISABLE_CHECK_COMPILER 1)
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply CMakeLists.txt:79]: That was the idea, but it seems not to work. Will remove.

File: src/disruptor/test/Rx_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    auto publisher = makePublisher(ringBuffer, input, patternRepeatCount);
+
+    // Rx events from the Disruptor
+    rx::Source<TestDisruptor> disruptorSource(testDisruptor);
+
+    // Splitting stream
+    auto equalTo = [](char device) {
+        return [=](const TestEvent &event) { return event.device == device; };
+    };
+    auto             deviceAStream  = disruptorSource.stream() | rxcpp::operators::filter(equalTo('a')) | rxcpp::operators::distinct_until_changed();
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply src/disruptor/test/Rx_tests.cpp:205]: It is not, I've added a comment that it is there just to demonstrate the transformation chaining.

File: src/disruptor/include/disruptor/Rx.hpp [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+
+                        if constexpr (EventCountLimitedPolicy<Policy>) {
+                            // If there is any queue with more than allowed items,
+                            // force sending
+                            if (!toSend) {
+                                detail::tuple_for_each(_observableBacklogs, [&forceSending]<typename IndexIn>(IndexIn, auto &observableBacklogIn) { forceSending |= observableBacklogIn.backlog.size() > Policy::eventCountLimit; });
+                            }
+                        }
+
+                        if constexpr (TimeLimitedPolicy<Policy>) {
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply src/disruptor/include/disruptor/Rx.hpp:331]: This is a passive timeout - that is, observed only if there are events from other streams.
I'll add the heartbeat option.

File: src/disruptor/test/CMakeLists.txt [hunk 5]
```diff
@@ -1,39 +1,48 @@
...
+        ${name}
+        PUBLIC
+            ${CMAKE_CURRENT_SOURCE_DIR}/../include
+            ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/include
+            ${CMAKE_CURRENT_SOURCE_DIR}/../../core/include
             ${CMAKE_CURRENT_SOURCE_DIR}/../../disruptor/include
-    )
+        )
+
+    target_include_directories (${name} SYSTEM PUBLIC ${rx_SOURCE_DIR}/Rx/v2/src)
```
[the remark this answers is not in the document]
Comment (ivan-cukic) [reply src/disruptor/test/CMakeLists.txt:28]: Done. I'm not that fond of adding header-only things to target_link_libraries, but that /is/ the modern cmake way :)

**Review by wirew0rm — APPROVED** [review]
Great. Thanks for addressing the issues, looks really nice now!

**Comment by wirew0rm** [comment]
restyled seems to be down, but I applied clang-format locally.

**Pull request merged**